### PR TITLE
Mark `Assembly.GetCallingAssembly` with `[RequiresDynamicCode]`.

### LINKED ISSUE #53825: [wasm][aot] Figure out how to handle APIs not supported on AOT
Some APIs, like `Assembly.GetCallingAssembly()` are not supported on wasm/AOT. We should have a clear behavior for this, eg. throw an exception, or maybe mark these APIs, so that the user is aware that these are not supported.

Example case: https://github.com/dotnet/runtime/issues/53815

How is this handled for non-wasm+aot cases?

cc @vargaz @steveisok

### LINKED ISSUE #71290: `Assembly.GetCallingAssembly` should be marked with `[RequiresDynamicCode]`.
As indicated in #53825, `Assembly.GetCallingAssembly` is not supported in AOT environments. It should be applied the `[RequiresDynamicCodeAttribute]` to make this visible.

### PR DESCRIPTION
**Pull request by teo-tsirpanis**
Fixes #71290
Fixes #53825

I added `[RequiresDynamicCode]` to `Assembly.GetDynamicAssembly` across all runtimes, and also marked all APIs that use it. The only uses of it in non-test code are `AssemblyBuilder.DefineDynamicAssembly` which obviously is already marked, and some APIs in `Microsoft.VisualBasic.Core` (c.c. @cston).

### CODE REVIEW

File: src/libraries/Microsoft.VisualBasic.Core/ref/Microsoft.VisualBasic.Core.cs [hunk 1]
```diff
@@ -329,6 +330,7 @@ public enum FileAttribute
         Directory = 16,
         Archive = 32,
     }
+    [System.Diagnostics.CodeAnalysis.RequiresDynamicCodeAttribute("The FileSystem module is not supported in AOT environments. Use members of the System.IO namespace instead.")]
     [Microsoft.VisualBasic.CompilerServices.StandardModuleAttribute]
     public sealed partial class FileSystem
```
Comment (teo-tsirpanis) [inline src/libraries/Microsoft.VisualBasic.Core/ref/Microsoft.VisualBasic.Core.cs:335]: Many but not all APIs of this class use `GetCallingAssembly` (some did call it only in debug asserts) but I didn't bother and marked the whole class. This class is very outdated either way.

**Review by jkotas — APPROVED** [review]

**Comment by MichalStrehovsky** [comment]
I wanted to comment when I saw the original issue but I forgot. I don't know if we have the right way to annotate this.

`RequiresDynamicCode` means we need to generate native code at runtime.

The problem here is different:

* On WASM, we are not allowed to do stack walk because the platform forbids it. We don't know who's the caller because of that.
* On NativeAOT, we can strip metadata about code that doesn't need it. We can do a stack walk, but we still might not know what `MethodBase` the code is associated with. It's kind of a trimming problem. It's just the trimming can be a lot more aggressive for native code. The more appropriate would be RequiresUnreferencedCode, but even that one is not exact.

I don't know what the right solution is. I'm just having a bit doubts about this one.

**Comment by jkotas** [comment]
The next best option would be to mark this API obsolete. https://github.com/dotnet/csharplang/issues/4984 would be required for that to have viable replacement.

**Comment by buyaa-n** [comment]
Sounds like could use combination of `RequiresUnreferencedCode` and `UnsupportedOSPlatform("browser")`

**Comment by MichalStrehovsky** [comment]
> Sounds like could use combination of `RequiresUnreferencedCode` and `UnsupportedOSPlatform("browser")`

Unfortunately WASM not supporting stackwalks is more nuanced - it works when we're running interpreter because we're not using the WASM execution stack. It doesn't work when the code was AOT'd.

I would really love it if we could just obsolete this API and not have this problem :/.

**Comment by buyaa-n** [comment]
Thanks all, as we don't want to fix the issue this way closing the PR

**Pull request closed without merging**